src/tike/pool.py

```diff
@@ -0,0 +1,85 @@
+"""Defines a worker Pool for multi-device managerment."""
+
+__author__ = "Daniel Ching"
+__copyright__ = "Copyright (c) 2020, UChicago Argonne, LLC."
+__docformat__ = 'restructuredtext en'
+__all__ = ['ThreadPool']
+
+from concurrent.futures import ThreadPoolExecutor
+import os
+
+import numpy as np
+
+
+class NumPyThreadPool(ThreadPoolExecutor):
+    """Python thread pool plus scatter gather methods.
+
+    A Pool is a context manager which provides access to and communications
+    amongst workers.
+
+    """
+
+    def __init__(self, num_workers: int):
+        super().__init__(num_workers)
+        self.num_workers = num_workers
+        self.workers = list(range(num_workers))
+        self.xp = np
+
+    def _copy_to(self, x: np.array, worker: int) -> np.array:
+        """Copy x to the given worker."""
+        return self.xp.array(x, copy=True)
+
+    def bcast(self, x: np.array) -> list:
+        """Send a copy of x to all workers."""
+
+        def f(worker):
+            return self._copy_to(x, worker)
+
+        return list(self.map(f, self.workers))
+
+    # def scatter(self, x: np.array) -> list:
+    #     """Divide x amongst all workers along the 0th dimension."""
+    #     return list(self.map(self._copy_to, x, self.workers))
+
+    def gather(self, x: list, worker=0, axis=0) -> np.array:
+        """Concatenate x on a single worker along the given axis."""
+        return self.xp.concatenate(
+            [self._copy_to(part, worker) for part in x],
+            axis,
+        )
+
+    def all_gather(self, x: list) -> list:
+        """Copy a scattered x to all workers."""
```

Comment:
Here is confusing. A scattered x should be already in all workers. Do you mean "Copy a gathered x to all workers."? Then, in this case, the function name should be `all_reduce`.

Reply:
Good eye! I meant `all_gather`; all threads gather x. `all_reduce` would involve taking a sum along some dimension of the array.

However, I guess the docstring should be:
```suggestion
        """Concatenate x on all worker along the given axis."""
```
and I should add the axis parameter to `all_gather`.